tests/vm_test.go

```diff
@@ -1921,6 +1924,65 @@ status:
 			}, 2*time.Minute, 1*time.Second).Should(Succeed())
 		})
 	})
+
+	FContext("[Serial] when node becomes unhealthy", Serial, func() {
+		const componentName = "virt-handler"
+
+		getHandlerNodePod := func(nodeName string) *k8sv1.PodList {
+			pods, err := virtClient.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(context.Background(),
+				metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("kubevirt.io=%s", componentName),
+					FieldSelector: fmt.Sprintf("spec.nodeName=" + nodeName),
+				})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(pods.Items).To(HaveLen(1))
+
+			return pods
+		}
+
+		It("[Serial] the VMs running in that node should be respawned", func() {
+
+			By("Starting VM")
+			vm := startVM(virtClient, createVM(virtClient, libvmi.NewAlpine()))
+			vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			nodeName := vmi.Status.NodeName
+			oldUID := vmi.UID
+
+			By("Blocking virt-handler for updating the VMI")
+			libpod.AddServiceBackHoleForKubernetesApiToPod(getHandlerNodePod(nodeName), virtClient, componentName)
+			Eventually(getHandlerNodePod(nodeName).Items[0], 120*time.Second, time.Second, HaveConditionFalse(k8sv1.PodReady))
+
+			DeferCleanup(func() {
+				libpod.DeleteServiceBlackHoleForKubernetesApiToPod(getHandlerNodePod(nodeName), virtClient, componentName)
+				Eventually(getHandlerNodePod(nodeName).Items[0], 120*time.Second, time.Second, HaveConditionTrue(k8sv1.PodReady))
+			})
+
+			pod, err := libvmi.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Simulating lost of the virt-launcher")
+			err = virtClient.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: ptr.To(int64(0)),
+			})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("The controller should recreate the launcher pod and move the VMI to phase READY")
+			Eventually(ThisVMI(vmi), 60*time.Second, 2*time.Second).Should(HaveConditionTrue(v1.VirtualMachineInstanceReady))
```

Comment:
How long will the VMI have a condition other than Ready? E.g. going from Ready -> !Ready -> Ready, would it make sense to assert it went to something other than Ready first?

Reply:
Instead of observing the VMI we could instead wait for the VM to become not ready? And then become ready again?

Reply:
I've added a check to test if the VMI is !Ready and if the VMI is on Scheduled state at some point.